jboss-eap-cd-jgroups/added/launch/ha.sh

```diff
@@ -0,0 +1,95 @@
+        if [ -f "${CA_CERT}" ]; then
+            CURL_CERT_OPTION="--cacert ${CA_CERT}"
+        else
+            CURL_CERT_OPTION="-k"
+        fi
+        pods_auth="Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)"
+        pods_code=$(curl --noproxy "*" -s -o /dev/null -w "%{http_code}" -G --data-urlencode "${pods_labels}" ${CURL_CERT_OPTION} -H "${pods_auth}" ${pods_url})
+        if [ "${pods_code}" = "200" ]; then
+            log_info "Service account has sufficient permissions to view pods in kubernetes (HTTP ${pods_code}). Clustering will be available."
+        elif [ "${pods_code}" = "403" ]; then
+            log_warning "Service account has insufficient permissions to view pods in kubernetes (HTTP ${pods_code}). Clustering might be unavailable. Please refer to the documentation for configuration."
+        else
+            log_warning "Service account unable to test permissions to view pods in kubernetes (HTTP ${pods_code}). Clustering might be unavailable. Please refer to the documentation for configuration."
+        fi
+    else
+        log_warning "Environment variable OPENSHIFT_KUBE_PING_NAMESPACE undefined. Clustering will be unavailable. Please refer to the documentation for configuration."
+    fi
+}
+
+function validate_dns_ping_settings() {
+  if [ "x$OPENSHIFT_DNS_PING_SERVICE_NAME" = "x" ]; then
+    log_warning "Environment variable OPENSHIFT_DNS_PING_SERVICE_NAME undefined. Clustering will be unavailable. Please refer to the documentation for configuration."
+  fi
+}
+
+function validate_ping_protocol() {
+  if [ "$1" = "openshift.KUBE_PING" ]; then
+    check_view_pods_permission
+  elif [ "$1" = "openshift.DNS_PING" ]; then
+    validate_dns_ping_settings
+  else
+    log_warning "Unknown protocol specified for JGroups discovery protocol: $1.  Expecting one of: openshift.KUBE_PING or openshift.DNS_PING."
+  fi
+}
+
+function configure_ha() {
+  # Set HA args
+  IP_ADDR=`hostname -i`
+  JBOSS_HA_ARGS="-b ${IP_ADDR} -bprivate ${IP_ADDR}"
+
+  init_node_name
+
+  JBOSS_HA_ARGS="${JBOSS_HA_ARGS} -Djboss.node.name=${JBOSS_NODE_NAME}"
+
+  if [ -z "${JGROUPS_CLUSTER_PASSWORD}" ]; then
+      log_warning "No password defined for JGroups cluster. AUTH protocol will be disabled. Please define JGROUPS_CLUSTER_PASSWORD."
+      JGROUPS_AUTH="<!--WARNING: No password defined for JGroups cluster. AUTH protocol has been disabled. Please define JGROUPS_CLUSTER_PASSWORD. -->"
+  else
+    JGROUPS_AUTH="\n\
+                <auth-protocol type=\"AUTH\">\n\
+                    <digest-token algorithm=\"SHA-512\">\n\
+                        <shared-secret-reference clear-text=\"$JGROUPS_CLUSTER_PASSWORD\"/>\n\
+                    </digest-token>\n\
+                </auth-protocol>\n"
+  fi
+
+  local ping_protocol=${JGROUPS_PING_PROTOCOL:-openshift.KUBE_PING}
+  local ping_protocol_element="<socket-discovery-protocol type=\"${ping_protocol}\"/>"
```

Comment:
This should just use &lt;protocol type="..."/&gt;.  &lt;socket-discovery-protocol/&gt; is meant to require a list of outbound socket bindings (for use by protocols like TCPPING and TCPGOSSIP).  Looks like our subsystem parser isn't properly validating the required attribute.  I've opened https://issues.jboss.org/browse/WFLY-10893 to address that.  Once that is fixed validation of the above will fail.

Reply:
Thanks @pferraro, will update to fix.